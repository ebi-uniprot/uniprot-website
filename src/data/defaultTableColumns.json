["accession", "id", "protein_name", "gene_names", "organism"]
