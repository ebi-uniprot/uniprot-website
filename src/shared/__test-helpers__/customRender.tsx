/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
