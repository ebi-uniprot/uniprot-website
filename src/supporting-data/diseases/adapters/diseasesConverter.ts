import { Statistics } from '../../../shared/types/apiModel';
import { KeywordNameID } from '../../keywords/adapters/keywordsConverter';

type XRef = {
  databaseType: string;
  properties?: string[];
  id: string;
};

export type DiseasesAPIModel = {
  keywords?: KeywordNameID[];
  alternativeNames?: string[];
  acronym: string;
  crossReferences: XRef[];
  name: string;
  id: string;
  definition: string;
  statistics?: Statistics;
};

export type DiseasesUIModel = DiseasesAPIModel & {
  // any addition/change by the converter
};

const diseasesConverter = (data: DiseasesAPIModel): DiseasesUIModel => ({
  ...data,
});

export default diseasesConverter;
