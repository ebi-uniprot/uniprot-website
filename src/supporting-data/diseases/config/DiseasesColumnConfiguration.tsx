import { Link } from 'react-router-dom';
import { ExpandableList } from 'franklin-sites';

import { getEntryPathFor } from '../../../app/config/urls';
import { mapToLinks } from '../../../shared/components/MapTo';

import { DiseasesAPIModel } from '../adapters/diseasesConverter';
import { ColumnConfiguration } from '../../../shared/types/columnConfiguration';
import { Namespace } from '../../../shared/types/namespaces';

import helper from '../../../shared/styles/helper.module.scss';

export enum DiseasesColumn {
  acronym = 'acronym',
  alternativeNames = 'alternative_names',
  crossReferences = 'cross_references',
  definition = 'definition',
  id = 'id',
  keywords = 'keywords',
  name = 'name',
  // TODO: remove this once the backend is fixed https://www.ebi.ac.uk/panda/jira/browse/TRM-26601
  reviewedProteinCount = 'reviewed_protein_count',
  // TODO: remove this once the backend is fixed https://www.ebi.ac.uk/panda/jira/browse/TRM-26601
  unreviewedProteinCount = 'unreviewed_protein_count',
  // NOTE: once the backend is fixed, this will be available https://www.ebi.ac.uk/panda/jira/browse/TRM-26601
  statistics = 'statistics',
}

export const defaultColumns = [
  DiseasesColumn.id,
  DiseasesColumn.name,
  DiseasesColumn.keywords,
];

export const primaryKeyColumns = [DiseasesColumn.id];

const getEntryPath = getEntryPathFor(Namespace.diseases);
const getEntryPathForKeyword = getEntryPathFor(Namespace.keywords);

export const DiseasesColumnConfiguration: ColumnConfiguration<
  DiseasesColumn,
  Partial<DiseasesAPIModel>
> = new Map();

// COLUMN RENDERERS BELOW
DiseasesColumnConfiguration.set(DiseasesColumn.acronym, {
  label: 'Acronym',
  render: ({ acronym }) => acronym,
});

DiseasesColumnConfiguration.set(DiseasesColumn.alternativeNames, {
  label: 'Alternative names',
  render: ({ alternativeNames }) => (
    <ExpandableList
      descriptionString="alternative names"
      displayNumberOfHiddenItems
    >
      {alternativeNames}
    </ExpandableList>
  ),
});

// NOTE: should probably be links
DiseasesColumnConfiguration.set(DiseasesColumn.crossReferences, {
  label: 'Cross references',
  // TODO: https://www.ebi.ac.uk/panda/jira/browse/TRM-25838
  render: ({ crossReferences }) => (
    <ExpandableList
      descriptionString="cross references"
      displayNumberOfHiddenItems
    >
      {crossReferences?.map(
        ({ databaseType, id, properties }) =>
          `${databaseType}: ${id}${
            properties?.length ? ` (${properties.join(', ')})` : ''
          }`
      )}
    </ExpandableList>
  ),
});

DiseasesColumnConfiguration.set(DiseasesColumn.definition, {
  label: 'Definition',
  render: ({ definition }) => definition,
});

DiseasesColumnConfiguration.set(DiseasesColumn.id, {
  label: 'ID',
  render: ({ id }) =>
    id && (
      <Link className={helper['no-wrap']} to={getEntryPath(id)}>
        {id}
      </Link>
    ),
});

DiseasesColumnConfiguration.set(DiseasesColumn.keywords, {
  label: 'Keywords',
  render: ({ keywords }) => (
    <ExpandableList descriptionString="keywords" displayNumberOfHiddenItems>
      {keywords?.map(({ name, id }) => (
        <Link key={id} to={getEntryPathForKeyword(id)}>
          {name}
        </Link>
      ))}
    </ExpandableList>
  ),
});

DiseasesColumnConfiguration.set(DiseasesColumn.name, {
  label: 'Name',
  render: ({ name }) => name,
});

DiseasesColumnConfiguration.set(DiseasesColumn.statistics, {
  label: 'Statistics',
  render: ({ id, statistics }) => (
    <ExpandableList>
      {mapToLinks(Namespace.diseases, id, statistics)?.map(
        ({ key, link, name }) => (
          // eslint-disable-next-line uniprot-website/use-config-location
          <Link key={key} to={link}>
            {name}
          </Link>
        )
      )}
    </ExpandableList>
  ),
});

export default DiseasesColumnConfiguration;
