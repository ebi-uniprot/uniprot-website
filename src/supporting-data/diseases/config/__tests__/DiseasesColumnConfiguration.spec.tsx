import DiseasesColumnConfiguration from '../DiseasesColumnConfiguration';

import diseasesConverter, {
  DiseasesAPIModel,
  DiseasesUIModel,
} from '../../adapters/diseasesConverter';
import customRender from '../../../../shared/__test-helpers__/customRender';

import data from '../../__mocks__/diseasesModelData';

jest.mock('../../../../tools/utils/storage');

describe('DiseasesColumnConfiguration component', () => {
  let transformedData: DiseasesUIModel;

  beforeAll(() => {
    transformedData = diseasesConverter(data[0] as DiseasesAPIModel);
  });

  // TODO: find mock data to create non-null unreviewed_protein_count snapshot
  test.each(Array.from(DiseasesColumnConfiguration.entries()))(
    `should render column "%s"`,
    (key, column) => {
      const { asFragment } = customRender(
        <>{column.render(transformedData)}</>
      );
      expect(asFragment()).toMatchSnapshot(key);
    }
  );
});
