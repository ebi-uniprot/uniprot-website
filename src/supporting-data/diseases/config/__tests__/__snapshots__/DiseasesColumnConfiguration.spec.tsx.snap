// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiseasesColumnConfiguration component should render column "acronym": acronym 1`] = `
<DocumentFragment>
  HNPCC1
</DocumentFragment>
`;

exports[`DiseasesColumnConfiguration component should render column "alternative_names": alternative_names 1`] = `
<DocumentFragment>
  <ul
    class="expandable-list no-bullet"
  >
    <li>
      Hereditary non-polyposis colorectal cancer 3
    </li>
    <li>
      HNPCC3
    </li>
    <li>
      Lynch cancer family syndrome
    </li>
    <li>
      Lynch syndrome
    </li>
    <li>
      Lynch syndrome type I
    </li>
    <li>
      Lynch syndrome type II
    </li>
  </ul>
</DocumentFragment>
`;

exports[`DiseasesColumnConfiguration component should render column "cross_references": cross_references 1`] = `
<DocumentFragment>
  <ul
    class="expandable-list no-bullet"
  >
    <li>
      MIM: 120435 (phenotype)
    </li>
    <li>
      MedGen: C0009405
    </li>
    <li>
      MedGen: C1333990
    </li>
    <li>
      MedGen: CN068508
    </li>
    <li>
      MeSH: D003123
    </li>
  </ul>
</DocumentFragment>
`;

exports[`DiseasesColumnConfiguration component should render column "definition": definition 1`] = `
<DocumentFragment>
  An autosomal dominant disease associated with marked increase in cancer susceptibility. It is characterized by a familial predisposition to early-onset colorectal carcinoma (CRC) and extra- colonic tumors of the gastrointestinal, urological and female reproductive tracts. HNPCC is reported to be the most common form of inherited colorectal cancer in the Western world. Clinically, HNPCC is often divided into two subgroups. Type I is characterized by hereditary predisposition to colorectal cancer, a young age of onset, and carcinoma observed in the proximal colon. Type II is characterized by increased risk for cancers in certain tissues such as the uterus, ovary, breast, stomach, small intestine, skin, and larynx in addition to the colon. Diagnosis of classical HNPCC is based on the Amsterdam criteria: 3 or more relatives affected by colorectal cancer, one a first degree relative of the other two; 2 or more generation affected; 1 or more colorectal cancers presenting before 50 years of age; exclusion of hereditary polyposis syndromes. The term 'suspected HNPCC' or 'incomplete HNPCC' can be used to describe families who do not or only partially fulfill the Amsterdam criteria, but in whom a genetic basis for colon cancer is strongly suspected.
</DocumentFragment>
`;

exports[`DiseasesColumnConfiguration component should render column "id": id 1`] = `
<DocumentFragment>
  <a
    class="no-wrap"
    href="/diseases/DI-00550"
  >
    DI-00550
  </a>
</DocumentFragment>
`;

exports[`DiseasesColumnConfiguration component should render column "keywords": keywords 1`] = `
<DocumentFragment>
  <ul
    class="expandable-list no-bullet"
  >
    <li>
      <a
        href="/keywords/KW-0362"
      >
        Hereditary nonpolyposis colorectal cancer
      </a>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`DiseasesColumnConfiguration component should render column "name": name 1`] = `
<DocumentFragment>
  Hereditary non-polyposis colorectal cancer 1
</DocumentFragment>
`;
